pallets/oracle-data-collection/src/tests.rs

```diff
@@ -0,0 +1,309 @@
+use cfg_traits::data::DataRegistry;
+use frame_support::{assert_err, assert_ok, storage::bounded_vec::BoundedVec};
+use sp_runtime::{testing::H256, traits::Get, DispatchError};
+
+use crate::{
+	mock::*,
+	pallet::{Config, Error, Keys},
+	types::Change,
+};
+
+const ADMIN: AccountId = 1;
+const ANY: AccountId = 100;
+const FEEDER_1: AccountId = 10;
+const FEEDER_2: AccountId = 11;
+const FEEDER_3: AccountId = 12;
+
+const COLLECTION_ID: CollectionId = 1;
+const KEY_A: OracleKey = 1;
+const KEY_B: OracleKey = 2;
+const KEY_ERR: OracleKey = 3;
+const KEY_NONE: OracleKey = 4;
+const CHANGE_ID: ChangeId = H256::repeat_byte(0x42);
+
+mod mock {
+	use super::*;
+
+	pub fn prepare_update_feeders(
+		key: OracleKey,
+		feeders: &BoundedVec<AccountId, MaxFeedersPerKey>,
+	) {
+		MockChangeGuard::mock_note({
+			let feeders = feeders.clone();
+			move |pool_id, change| {
+				assert_eq!(pool_id, COLLECTION_ID);
+				assert_eq!(change, Change::Feeders(key, feeders.clone()));
+				Ok(CHANGE_ID)
+			}
+		});
+		MockChangeGuard::mock_released({
+			let feeders = feeders.clone();
+			move |pool_id, change_id| {
+				assert_eq!(pool_id, COLLECTION_ID);
+				assert_eq!(change_id, CHANGE_ID);
+				Ok(Change::Feeders(key, feeders.clone()))
+			}
+		});
+		MockIsAdmin::mock_check(|(admin, collection_id)| {
+			assert_eq!(admin, ADMIN);
+			assert_eq!(collection_id, COLLECTION_ID);
+			true
+		});
+	}
+
+	pub fn prepare_provider() {
+		MockProvider::mock_get(|(account, collection_id), key| {
+			assert_eq!(collection_id, &COLLECTION_ID);
+			match (account, key) {
+				(&FEEDER_1, &KEY_A) => Ok(Some((100, 50))),
+				(&FEEDER_2, &KEY_A) => Ok(Some((101, 45))),
+				(&FEEDER_3, &KEY_A) => Ok(Some((102, 55))),
+				(&FEEDER_1, &KEY_B) => Ok(Some((1000, 500))),
```

Comment:
Nit: Since these values are used throughout the test, it would be nice to define them as consts as well to improve readability.